sessions/2018/october/Session_15.md

```diff
@@ -0,0 +1,324 @@
+
+    BookController(BookService bookService) {
+        this.bookService = bookService;
+    }
+
+    // ...
+}
+```
+
+__NOTE__: Constructor injection is the best choice to use in main code.
+
+__NOTE__: When we declare a field as final, it cannot be modified after, thus we have quicker applications and less prone to modifications by mistake in the fields we don't want. By other side, is not recommended use final keyword in classes and methods because that lead to applications that are hard to work with.
+
+#### Testing in Spring
+
+When we are testing services in Spring, we can built separate test files and if we are testing services that have not dependencies on Spring components we can use test in normal way without use `@Autowire` annotation (injection). For example:
+
+```java
+package co.org.osso.api;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ApiApplicationBookTests {
+
+    private BookService BookServiceTarget = new BookService();
+
+    @Test
+    public void checkBookServiceGetBookWhenNull() {
+        Book book = BookServiceTarget.getBook(0);
+        Assert.assertEquals(null, book.getBookID());
+    }
+
+}
+```
+
+But if we have dependencies, for example __when testing a controller__, so we must to use `@Autowire` to link dependencies and call controller's methods directly, not by hitting the URLs of HTTP methods (GET, POST, PUT, etc.) in the controller maps. For example:
+
+```java
+package co.org.osso.api;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class ApiApplicationBookTests {
+
+    private BookService BookServiceTarget = new BookService();
+
+    @Autowired
+    private BookController BookControllerTarget;
+
+    @Test
+    public void checkBookServiceGetBookWhenNull() {
+        Book book = BookServiceTarget.getBook(0);
+        Assert.assertEquals(null, book.getBookID());
+    }
+
+    @Test
+    public void checkBookControllerGetBookEqualsTo() {
+      Book book = new Book();
+      book.setTitle("Book in Test");
+      book.setYear("1900");
+      BookControllerTarget.postBook(book);
+      Assert.assertEquals("1900", BookControllerTarget.getBook(0).getYear());
+    }
+
+}
+```
+
+Something to notice in the previous example is that I had to create a new book from the test self, If I don't do so, this will return `null` even if I already have populated the in-memory array from Postman. The reason for this is that my test is running in a different memory space than the running application, just like when I'm running a Java application twice or when I when I'm using two different linux terminals. This is for temporal in-memory datasets, if I'm using a database, data will be persistent.
```

Comment:
Yup! The memory spaces are different because you're actually running two different processes. Just like opening up two terminals at once. The output of one terminal isn't shown in the other.